# Cell Ranger

Cell Ranger is a software developed by 10x Genomics for processing and analyzing data from single-cell RNA sequencing (scRNA-seq) experiments.

Once the FASTQ files are generated by the sequencer, Cell Ranger aligns the sequencing reads to a reference genome. 

## Pipelines

- Cell ranger counts: gene expression 

- Cell ranger vdj: VDJ recpmbination in T/B cells. Identifies clonotypes and sequences of VDJ regions.

- Cell ranger multi: gene expression + VDJ + antibody capture. 

## Functionality: 

In single-cell experiments, each cell is tagged with a unique barcode. Cell Ranger identifies these barcodes to assign each read to its corresponding cell and it also counts the UMIs associated with each gene in each cell, providing a quantitative 
measure of gene expression ---> Cellr anger outputs a filtered gene expression matrix (h5), used for downstream analysis (Seurat). 


Cell Ranger also includes a specialized module called Cell Ranger V(D)J that is designed for processing and analyzing single-cell immune profiling data, specifically focused on the adaptive immune repertoire,
including T-cell and B-cell receptors. In this particular case, instead of aligning reads to the whole transcriptome, Cell Ranger V(D)J aligns them to a specialized reference that includes sequences for V (variable), D (diversity), J (joining),
and C (constant) gene segments of immune receptors. ---> Targeted algiment to VDJ sequences.
The outputs of cell ranger vdj and multi include: annotated V(D)J contigs, clonotype information, and summaries of V(D)J usage. The important ouput files are: Clonotypes.csv, filtered_contig_annotations.csv, and airr_rearangment.tsv.

If combined with gene expression profiling, Cell Ranger V(D)J can also provide information on the gene expression profile of the same cells.

