##### HBC single-cell RNA-seq workshop

# Single-cell RNA-seq analysis - QC

##########################################################

# Load libraries
library(SingleCellExperiment)
library(Seurat)
library(tidyverse)
library(Matrix)
library(scales)
library(cowplot)
library(RCurl)

## For a single sample:
# Read data generated by cell Ranger: 
ctrl_counts <- Read10X(data.dir = "data/ctrl_raw_feature_bc_matrix")

# Create a seurat object. This is for one sample.
ctrl <- CreateSeuratObject(counts = ctrl_counts,
                           min.features = 100)

head(ctrl@meta.data)

# ---------------------------

## For multiple samples:
# Create a Seurat object for multiples samples: 
for (file in c("ctrl_raw_feature_bc_matrix", "stim_raw_feature_bc_matrix")){
  seurat_data <- Read10X(data.dir = paste0("data/", file))
  seurat_obj <- CreateSeuratObject(counts = seurat_data, 
                                   min.features = 100, 
                                   project = file)
  assign(file, seurat_obj)
}
# For each sample, cell Ranger returns a folder with the barcode,features, and the matrix. 

# Check the metadata in the new Seurat objects
head(ctrl_raw_feature_bc_matrix@meta.data)
head(stim_raw_feature_bc_matrix@meta.data)

# Merge both Seurat objects into a single one. 
merged_seurat <- merge(x = ctrl_raw_feature_bc_matrix, 
                       y = stim_raw_feature_bc_matrix, 
                       add.cell.id = c("ctrl", "stim"))

# For more than 2 samples: 
# y = c(stim1_raw_feature_bc_matrix, stim2_raw_feature_bc_matrix, stim3_raw_feature_bc_matrix)
# add.cell.id = c("ctrl", "stim1", "stim2", "stim3")

# Because the same cell IDs can be used for different samples, we add a sample-specific prefix to each of our 
# cell IDs using the add.cell.id argument.

# Concatenate the count matrices of both samples together
merged_seurat <- JoinLayers(merged_seurat) 

head(merged_seurat@meta.data)
tail(merged_seurat@meta.data)


# ---------------------------------------------

# Quality Control: 

# Need to calculate additional metrics: 
# - Number of genes detected per UMI: this metric with give us an idea of the complexity of our dataset (more genes detected per UMI, more complex our data)
# - Mitochondrial ratio: give us a percentage of cell reads originating from the mitochondrial genes


# Genes per UMI: 
# log10(n.genes)/log10(n.UMI)

merged_seurat$log10GenesPerUMI <- log10(merged_seurat$nFeature_RNA) / log10(merged_seurat$nCount_RNA)


# Mitochondrial ratio: 
# Evaluates if there is mitochondrial contamination from dead cells.
# Poor quality cells --> ratio over 0.2 (or 20%)
merged_seurat$mitoRatio <- PercentageFeatureSet(object = merged_seurat, pattern = "^MT-") #this is a percentage
merged_seurat$mitoRatio <- merged_seurat@meta.data$mitoRatio / 100 #Fraction 

head(merged_seurat@meta.data)


# -------------------------------

## Additional metadata: 
metadata <- merged_seurat@meta.data

# Create a new column for cell identifiers: 
metadata$cells <- rownames(metadata)

# Create a new column indicating the condition of the sample: (depending on the prefix of the rownames)
metadata$sample <- NA
metadata$sample[which(str_detect(metadata$cells, "^ctrl_"))] <- "ctrl"
metadata$sample[which(str_detect(metadata$cells, "^stim_"))] <- "stim"

# Rename columns: 
metadata <- metadata %>%
  dplyr::rename(seq_folder = orig.ident,
                nUMI = nCount_RNA,
                nGene = nFeature_RNA)

# Save the metadata to the seurat object: 
merged_seurat@meta.data<-metadata

# Create .RData object to load at any time
save(merged_seurat, file="data/merged_filtered_seurat.RData")



# -------------------------------------------

# QC VISUALIZATION: 

# --- CELL COUNTS ---
# Visualize the number of cell counts per sample
metadata %>% 
  ggplot(aes(x=sample, fill=sample)) + 
  geom_bar() +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  theme(plot.title = element_text(hjust=0.5, face="bold")) +
  ggtitle("NCells")

# over 15,000 cells per sample, which is quite a bit more than the 12-13,000 expected. 
# It is clear that we likely have some junk ‘cells’ present.


# --- UMI (transcripts) ---
# Generally be above 500
# Visualize the number UMIs/transcripts per cell
metadata %>% 
  ggplot(aes(color=sample, x=nUMI, fill= sample)) + 
  geom_density(alpha = 0.2) + 
  scale_x_log10() + 
  theme_classic() +
  ylab("Cell density") +
  geom_vline(xintercept = 500)


# --- Genes per cell ---
# Visualize the distribution of genes detected per cell via histogram
metadata %>% 
  ggplot(aes(color=sample, x=nGene, fill= sample)) + 
  geom_density(alpha = 0.2) + 
  theme_classic() +
  scale_x_log10() + 
  geom_vline(xintercept = 300)


# --- Compelxity ---
# novelty score --> nGenes/nUMI
# Generally, we expect the novelty score to be above 0.80 for good quality cells.
# Visualize the overall complexity of the gene expression by visualizing the genes detected per UMI (novelty score)
metadata %>%
  ggplot(aes(x=log10GenesPerUMI, color = sample, fill=sample)) +
  geom_density(alpha = 0.2) +
  theme_classic() +
  geom_vline(xintercept = 0.8)


# --- Mitochondrial counts ratio ---
# Visualize the distribution of mitochondrial gene expression detected per cell
metadata %>% 
  ggplot(aes(color=sample, x=mitoRatio, fill=sample)) + 
  geom_density(alpha = 0.2) + 
  scale_x_log10() + 
  theme_classic() +
  geom_vline(xintercept = 0.2)


# --- Joint effects ---
# NGenes vs nUMIs colored by mitochondrial reads
# Visualize the correlation between genes detected and number of UMIs and determine whether strong presence of cells with low numbers of genes/UMIs
metadata %>% 
  ggplot(aes(x=nUMI, y=nGene, color=mitoRatio)) + 
  geom_point() + 
  scale_colour_gradient(low = "gray90", high = "black") +
  stat_smooth(method=lm) +
  scale_x_log10() + 
  scale_y_log10() + 
  theme_classic() +
  geom_vline(xintercept = 500) +
  geom_hline(yintercept = 250) +
  facet_wrap(~sample)
# Good cells--> high nGenes and high nUMIs


########################################################################
########################################################################

## FILTERING

# --- Cell level filtering ---

# Thresholds: 
# nUMI > 500
# nGene > 250
# log10GenesPerUMI > 0.8
# mitoRatio < 0.2

filtered_seurat <- subset(x = merged_seurat, 
                          subset= (nUMI >= 500) & 
                            (nGene >= 250) & 
                            (log10GenesPerUMI > 0.80) & 
                            (mitoRatio < 0.20))


# --- Gene level filtering ---

## Remove genes that have many 0 counts. 
# Extract counts
counts <- GetAssayData(object = filtered_seurat, layer = "counts")

# Output a logical matrix specifying for each gene on whether or not there are more than zero counts per cell
nonzero <- counts > 0

#keep only genes which are expressed in 10 or more cells.
keep_genes <- Matrix::rowSums(nonzero) >= 10
filtered_counts <- counts[keep_genes, ]


# Reassign to filtered Seurat object
filtered_seurat <- CreateSeuratObject(filtered_counts, meta.data = filtered_seurat@meta.data)

# Create .RData object to load at any time
save(filtered_seurat, file="data/seurat_filtered.RData")
